### PR TITLE
Spring-06

### PR DESCRIPTION
**Pull request by Max-Pochepets**
Further implementation of Spring MVC

### CODE REVIEW

**Review by Mandybur800 — APPROVED** [review]

File: src/main/java/cinema/shop/controller/AuthenticationController.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+import cinema.shop.model.dto.request.UserRequestDto;
+import cinema.shop.security.AuthenticationService;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/")
```
Comment (KseniiaMakarova) [inline src/main/java/cinema/shop/controller/AuthenticationController.java]: we can remove this

File: src/main/java/cinema/shop/controller/UserController.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+                          DtoResponseMapper<UserResponseDto, User> responseMapper) {
+        this.userService = userService;
+        this.responseMapper = responseMapper;
+    }
+
+    @GetMapping("/by-email")
+    public UserResponseDto findByEmail(@RequestParam String email) {
+        Optional<User> optionalUser = userService.findByEmail(email);
+        if (optionalUser.isEmpty()) {
+            return null;
```
Comment (KseniiaMakarova) [inline src/main/java/cinema/shop/controller/UserController.java]: let's better throw an exception

File: src/main/java/cinema/shop/model/dto/response/OrderResponseDto.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
+package cinema.shop.model.dto.response;
+
+import java.util.List;
+
+public class OrderResponseDto {
+    private Long id;
+    private List<Long> ticketsId;
```
Comment (KseniiaMakarova) [inline src/main/java/cinema/shop/model/dto/response/OrderResponseDto.java]:
```suggestion
    private List<Long> ticketIds;
```

File: src/main/java/cinema/shop/model/dto/response/ShoppingCartResponseDto.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
+package cinema.shop.model.dto.response;
+
+import java.util.List;
+
+public class ShoppingCartResponseDto {
+    private Long id;
+    private List<Long> ticketsId;
```
Comment (KseniiaMakarova) [inline src/main/java/cinema/shop/model/dto/response/ShoppingCartResponseDto.java]: same

File: src/main/java/cinema/shop/service/impl/MovieSessionServiceImpl.java [hunk 5]
```diff
@@ -25,13 +26,22 @@ public MovieSession add(MovieSession movieSession) {
         return movieSessionDao.add(movieSession);
     }
 
+    @Override
+    public MovieSession get(Long id) {
+        Optional<MovieSession> optionalMovieSession = movieSessionDao.get(id);
+        if (optionalMovieSession.isEmpty()) {
+            return null;
```
Comment (KseniiaMakarova) [inline src/main/java/cinema/shop/service/impl/MovieSessionServiceImpl.java]: let's throw an exception instead

File: src/main/java/cinema/shop/service/impl/MovieServiceImpl.java [hunk 6]
```diff
@@ -22,11 +22,8 @@ public Movie add(Movie movie) {
 
     @Override
     public Movie get(Long id) {
-        Optional<Movie> optionalMovie = movieDao.get(id);
-        if (optionalMovie.isEmpty()) {
-            return null;
-        }
-        return optionalMovie.get();
+        return movieDao.get(id).orElseThrow(()
+                -> new NoSuchElementException("There is no such movie."));
```
Comment (KseniiaMakarova) [inline src/main/java/cinema/shop/service/impl/MovieServiceImpl.java:26]: let's have the same approach for all the entities maybe?

File: src/main/java/cinema/shop/service/dto/mapping/impl/response/UserResponseMapper.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+@Component
+public class UserResponseMapper implements DtoResponseMapper<UserResponseDto, User> {
+    @Override
+    public UserResponseDto toDto(User object) {
+        UserResponseDto responseDto = new UserResponseDto();
+
+        responseDto.setId(object.getId());
+        responseDto.setEmail(object.getEmail());
+
```
Comment (KseniiaMakarova) [inline src/main/java/cinema/shop/service/dto/mapping/impl/response/UserResponseMapper.java]: remove redundant empty lines - the code is harder to read

File: src/main/java/cinema/shop/service/dto/mapping/impl/response/UserResponseMapper.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+import cinema.shop.model.User;
+import cinema.shop.model.dto.response.UserResponseDto;
+import cinema.shop.service.dto.mapping.DtoResponseMapper;
+import org.springframework.stereotype.Component;
+
+@Component
+public class UserResponseMapper implements DtoResponseMapper<UserResponseDto, User> {
+    @Override
+    public UserResponseDto toDto(User object) {
```
Comment (KseniiaMakarova) [inline src/main/java/cinema/shop/service/dto/mapping/impl/response/UserResponseMapper.java]:
```suggestion
    public UserResponseDto toDto(User user) {
```

File: src/main/java/cinema/shop/service/dto/mapping/impl/response/ShoppingCartResponseMapper.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+import cinema.shop.model.dto.response.ShoppingCartResponseDto;
+import cinema.shop.service.dto.mapping.DtoResponseMapper;
+import java.util.stream.Collectors;
+import org.springframework.stereotype.Component;
+
+@Component
+public class ShoppingCartResponseMapper
+        implements DtoResponseMapper<ShoppingCartResponseDto, ShoppingCart> {
+    @Override
+    public ShoppingCartResponseDto toDto(ShoppingCart object) {
```
Comment (KseniiaMakarova) [inline src/main/java/cinema/shop/service/dto/mapping/impl/response/ShoppingCartResponseMapper.java]: same

File: src/main/java/cinema/shop/service/dto/mapping/impl/response/ShoppingCartResponseMapper.java [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+        ShoppingCartResponseDto responseDto = new ShoppingCartResponseDto();
+
+        responseDto.setId(object.getId());
+        responseDto.setUserEmail(object.getUser().getEmail());
+        responseDto.getTicketsId().addAll(object.getTickets().stream()
```
Comment (KseniiaMakarova) [inline src/main/java/cinema/shop/service/dto/mapping/impl/response/ShoppingCartResponseMapper.java]:
```suggestion
        responseDto.setTicketIds(object.getTickets().stream()
```

File: src/main/java/cinema/shop/service/dto/mapping/impl/response/OrderResponseMapper.java [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+import org.springframework.stereotype.Component;
+
+@Component
+public class OrderResponseMapper implements DtoResponseMapper<OrderResponseDto, Order> {
+    @Override
+    public OrderResponseDto toDto(Order object) {
+        OrderResponseDto responseDto = new OrderResponseDto();
+
+        responseDto.setId(object.getId());
+        responseDto.getTicketsId().addAll(object.getTickets().stream()
```
Comment (KseniiaMakarova) [inline src/main/java/cinema/shop/service/dto/mapping/impl/response/OrderResponseMapper.java]: same comments here

File: src/main/java/cinema/shop/service/dto/mapping/impl/response/OrderResponseMapper.java [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+                                                .map(Ticket::getId)
+                                                .collect(Collectors.toList()));
+        responseDto.setOrderDate(object.getOrderDate().format(DateTimeFormatter
+                                                                .ofPattern("dd.MM.yyyy HH:mm")));
```
Comment (KseniiaMakarova) [inline src/main/java/cinema/shop/service/dto/mapping/impl/response/OrderResponseMapper.java]: move the pattern to a constant field

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/cinema/shop/controller/AuthenticationController.java [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+
+import cinema.shop.model.dto.request.UserRequestDto;
+import cinema.shop.security.AuthenticationService;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping
```
Comment (KseniiaMakarova) [inline src/main/java/cinema/shop/controller/AuthenticationController.java]: what is the purpose of this annotation here?

File: src/main/java/cinema/shop/service/dto/mapping/impl/response/OrderResponseMapper.java [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+import cinema.shop.model.Ticket;
+import cinema.shop.model.dto.response.OrderResponseDto;
+import cinema.shop.service.dto.mapping.DtoResponseMapper;
+import java.time.format.DateTimeFormatter;
+import java.util.stream.Collectors;
+import org.springframework.stereotype.Component;
+
+@Component
+public class OrderResponseMapper implements DtoResponseMapper<OrderResponseDto, Order> {
+
```
Comment (KseniiaMakarova) [inline src/main/java/cinema/shop/service/dto/mapping/impl/response/OrderResponseMapper.java]: redundant empty line

File: src/main/java/cinema/shop/service/impl/MovieSessionServiceImpl.java [hunk 15]
```diff
@@ -25,13 +26,19 @@ public MovieSession add(MovieSession movieSession) {
         return movieSessionDao.add(movieSession);
     }
 
+    @Override
+    public MovieSession get(Long id) {
+        return movieSessionDao.get(id).orElseThrow(()
+                -> new NoSuchElementException("Could not get user by id " + id + "."));
```
Comment (KseniiaMakarova) [inline src/main/java/cinema/shop/service/impl/MovieSessionServiceImpl.java]: why user?

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — APPROVED** [review]
see a few comments above

**Pull request merged**